src/groth16/aggregate/accumulator.rs

```diff
@@ -60,58 +60,48 @@ impl<E: Engine, R: rand::RngCore + Send> PairingChecks<E, R> {
         self.valid.store(false, SeqCst);
     }
 
-    fn merge_pair(&self, result: E::Fqk, exp: E::Fqk) {
-        self.merge(PairingCheck::from_pair(result, exp));
+    fn merge_pair(&self, result: E::Fqk, exp: E::Fqk, must_randomize: bool) {
+        self.merge(PairingCheck::from_pair(result, exp), must_randomize);
     }
 
-    fn merge_miller_one(&self, result: E::Fqk) {
-        self.merge(PairingCheck::from_miller_one(result));
+    fn merge_miller_one(&self, result: E::Fqk, must_randomize: bool) {
+        self.merge(PairingCheck::from_miller_one(result), must_randomize);
     }
 
-    pub fn merge_parts<F, G, H, I>(&self, f: F, g: G, h: H, i: I)
-    where
-        F: Fn() -> E::Fqk + Send,
-        G: Fn() -> E::Fqk + Send,
-        H: Fn() -> E::Fqk + Send,
-        I: Fn() -> (E::Fqk, E::Fqk) + Send,
-    {
-        assert!(
-            !self.non_random_check_done.load(SeqCst),
-            "merge_parts can only be called once"
-        );
+    pub fn merge_pairing_equation(&self, left: Vec<E::Fqk>, pair: (E::Fqk, E::Fqk)) {
+        let must_randomize = self.non_random_check_done.load(SeqCst);
 
-        rayon::scope(|s| {
-            s.spawn(move |_| {
-                self.merge_miller_one(f());
-            });
-            s.spawn(move |_| {
-                self.merge_miller_one(g());
-            });
-            s.spawn(move |_| {
-                self.merge_miller_one(h());
-            });
-            s.spawn(move |_| {
-                let (a, b) = i();
-                self.merge_pair(a, b);
-            });
-        });
-
-        self.non_random_check_done.store(true, SeqCst);
+        for l in left.iter() {
+            self.merge_miller_one(*l, must_randomize);
```

Comment:
i guess it's not a long way to get an array of closures and iter in parallel if we wanted to put back parallelization.